frame/transaction-storage/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+use sp_transaction_storage_proof::{
+	TransactionStorageProof, InherentError,
+	random_chunk, encode_index,
+	CHUNK_SIZE, INHERENT_IDENTIFIER, DEFAULT_STORAGE_PERIOD,
+};
+
+/// A type alias for the balance type from this pallet's point of view.
+type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// Maximum bytes that can be storead in one transaction.
+// Increasing it further also requires raising the allocator limit.
+pub const MAX_DATA_SIZE: u32 = 8 * 1024 * 1024;
+
+/// State data for a stored transaction.
+#[derive(Encode, Decode, sp_runtime::RuntimeDebug, PartialEq, Eq)]
+pub struct TransactionInfo {
+	/// Chunk trie root.
+	chunk_root: <BlakeTwo256 as Hash>::Output,
+	/// Plain hash of indexed data.
+	content_hash: <BlakeTwo256 as Hash>::Output,
+	/// Size of indexed data in bytes.
+	size: u32,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// A dispatchable call.
+		type Call: Parameter + Dispatchable<Origin=Self::Origin> + GetDispatchInfo + From<frame_system::Call<Self>>;
+		/// The currency trait.
+		type Currency: ReservableCurrency<Self::AccountId>;
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Insufficient account balance.
+		InsufficientFunds,
+		/// Invalid configuration.
+		NotConfigured,
+		/// Renewed extrinsic is not found.
+		RenewedNotFound,
+		/// Attempting to store empty transaction
+		EmptyTransaction,
+		/// Proof was not expected in this block.
+		UnexpectedProof,
+		/// Proof failed verification.
+		InvalidProof,
+		/// Missing storage proof.
+		MissingProof,
+		/// Unable to verify proof becasue state data is missing.
+		MissingStateData,
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			let transaction_count = if obsolete > Zero::zero() {
+				<TransactionCount<T>>::get(obsolete)
+			} else {
+				0
+			};
+			// return weight for `on_finalize` does
+			T::WeightInfo::on_finalize(transaction_count)
+		}
+
+		fn on_finalize(n: T::BlockNumber) {
+			<TransactionCount<T>>::insert(n, <Counter<T>>::get());
+			<Counter<T>>::kill();
+			// Drop obsolete roots
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			if obsolete > Zero::zero() {
```

Comment:
should use a checked sub with a match statement.

This will clean up things immediately for the whole first period of a chain, where n is less than period.

Reply:
Not sure what you mean here. `saturating_sub` returns 0 when underflows, so strict check on the following line won't trigger the cleanup. We also don't want to cleanup block 0.

Reply:
ah you are right since obsolete must be greater than zero, not zero itself.